src/charm.py

```diff
@@ -115,9 +114,24 @@ def on_leader_elected(self, _):
 
     def on_database_joined(self, event):
         """Run the joined hook for the database relation."""
+        if not self.unit.is_leader():
+            return
+
+        # Set address and port
         self.provider.update_port("database", CQL_PORT)
         self.provider.update_address("database", self._hostname())
 
+        # Set credentials
+        creds = self.provider.credentials(event.relation.id)
+        if not creds:
+            # Create a user for the related charm to use.
+            username = f"juju-user-{event.app.name}"
```

Comment:
I have a vague memory of discussing uniqueness of the app name. Not a concern here?
```suggestion
            username = f"juju-user-{event.relation.id}-{event.app.name}"
```

Reply:
App names looked up like this are effectively unique, but in CMR cases opaque: within the same model, the app name is unique. Across models, you see those `remote-<random_stuff>` ids.

Also, this sounds like a good integration test.

Reply:
@mmanciop I thought this actually would be an issue. Are you saying that event.app.name will have stuff appended automatically?

Reply:
Not necessarily. The app name will match what Juju shows, unless it's a cross-model relation, in which case it will be `remote-$(hex string)`, and not the actual app name. 

This is still unique (relative to the relation), and doesn't include any characters you won't see here, so it should be fine unless Cassandra has a limit on username length